### ISSUE TITLE
[full-determinism] Qwen_2_vl: training loss fails to converge to the same values with image inputs

### ISSUE THREAD

**Issue by xzzWZY** [ByteDance-Seed/VeOmni#540]
### System Info

Transformers: 4.57.3
torch: 2.9.1+cu129
NVIDIA driver: 535.161.08
nvidia-cudnn-cu12: 9.15.1.9

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

For Qwen 2 vl:

1. training losses aren't identical between two runs, with same setting for dataset with images
2. training losses are identical wit text-only dataset like tulu-3-sft-mixture
3. set  "torch.use_deterministic_algorithms(True, warn_only=False)" wouldn't lead to any errors

The detailed setting is shown here:

```
{
  "model": {
    "config_path": "../hf_cache/Qwen2-VL-7B-Instruct/",
    "model_path": "../hf_cache/Qwen2-VL-7B-Instruct/",
    "tokenizer_path": "../hf_cache/Qwen2-VL-7B-Instruct/",
    "safetensor_idx_path": "../hf_cache/Qwen2-VL-7B-Instruct/model.safetensors.index.json",
    "foundation": {},
    "encoders": {},
    "decoders": {},
    "input_encoder": "encoder",
    "output_encoder": "decoder",
    "encode_target": false,
    "attn_implementation": "veomni_flash_attention_2_with_sp",
    "moe_implementation": null,
    "basic_modules": [],
    "encoder_data_balance": false,
    "encoder_data_balance_sorting_algo": "post_mbs_balancing_greedy_without_pad"
  },
  "data": {
    "train_path": "../hf_cache/ShareGPT4V_dataset/sharegpt4v_instruct_gpt4-vision_cap100k_coco.json",
    "eval_path": null,
    "train_size": 80000000,
    "train_sample": 10000,
    "data_type": "conversation",
    "dataloader_type": "native",
    "datasets_type": "iterable",
    "multisource_datasets_type": "interleave",
    "source_name": "sharegpt4v_sft",
    "dyn_bsz_buffer_size": 200,
    "text_keys": "messages",
    "chat_template": "qwen2vl",
    "max_seq_len": 4096,
    "num_workers": 2,
    "prefetch_factor": 2,
    "drop_last": true,
    "pin_memory": true,
    "silent_exception": false,
    "mm_configs": {
      "image_max_pixels": 602112,
      "video_max_pixels": 602112,
      "max_frames": 16,
      "fps": 2.0,
      "use_audio_in_video": false
    }
  },
  "train": {
    "output_dir": "/tmp/qwen2_vl_transformer_4573",
    "train_architecture": "full",
    "dyn_bsz": true,
    "lr": 1e-05,
    "lr_min": 1e-07,
    "lr_start": 0.0,
    "weight_decay": 0,
    "no_decay_modules": [],
    "no_decay_params": [],
    "optimizer": "adamw",
    "max_grad_norm": 1.0,
    "micro_batch_size": 1,
    "global_batch_size": 16,
    "num_train_epochs": 1,
    "pad_to_length": false,
    "bsz_warmup_ratio": 0,
    "bsz_warmup_init_mbtoken": 200,
    "lr_warmup_ratio": 0,
    "lr_decay_style": "cosine",
    "lr_decay_ratio": 1.0,
    "enable_reshard_after_forward": true,
    "enable_reshard_after_backward": true,
    "enable_mixed_precision": true,
    "enable_gradient_checkpointing": true,
    "debug_gradient_checkpointing": false,
    "enable_reentrant": false,
    "enable_full_shard": true,
    "enable_forward_prefetch": true,
    "enable_fsdp_offload": false,
    "enable_activation_offload": false,
    "activation_gpu_limit": 0.0,
    "init_device": "meta",
    "broadcast_model_weights_from_rank0": true,
    "enable_full_determinism": true,
    "enable_batch_invariant_mode": false,
    "empty_cache_steps": 500,
    "gc_steps": 500,
    "data_parallel_mode": "fsdp2",
    "data_parallel_replicate_size": 1,
    "data_parallel_shard_size": 8,
    "tensor_parallel_size": 1,
    "expert_parallel_size": 1,
    "ep_outside": false,
    "pipeline_parallel_size": 1,
    "ulysses_parallel_size": 1,
    "async_enabled": false,
    "context_parallel_size": 1,
    "ckpt_manager": "dcp",
    "save_async": false,
    "load_checkpoint_path": null,
    "save_steps": 0,
    "save_epochs": 1,
    "hf_save_steps": 0,
    "hf_save_epochs": 0,
    "eval_steps": 0,
    "eval_epochs": 1,
    "save_hf_weights": false,
    "seed": 42,
    "enable_compile": false,
    "use_wandb": false,
    "wandb_project": "qwen3_vl_dense",
    "wandb_name": "qwen3_vl_dense",
    "wandb_id": null,
    "enable_profiling": false,
    "profile_start_step": 1,
    "profile_end_step": 2,
    "profile_trace_dir": "./trace",
    "profile_record_shapes": true,
    "profile_profile_memory": true,
    "profile_with_stack": true,
    "profile_rank0_only": true,
    "max_steps": 50,
    "freeze_vit": false,
    "freeze_audio_tower": false,
    "vit_lr": 1e-06
  }
}
```

### Expected behavior

each run with identical training losses

**Comment by Coach257** [comment]
use `VEOMNI_USE_LIGER_KERNEL=0` will enable fully deterministic.
